policy-controller/k8s/status/src/index.rs

```diff
@@ -0,0 +1,227 @@
+use crate::{
+    http_route::{ParentReference, RouteBinding},
+    resource_id::ResourceId,
+};
+use ahash::{AHashMap as HashMap, AHashSet as HashSet};
+use chrono::offset::Utc;
+use linkerd_policy_controller_k8s_api::{self as k8s, gateway, ResourceExt};
+use parking_lot::RwLock;
+use std::{collections::hash_map::Entry, sync::Arc};
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+
+const POLICY_API_GROUP: &str = "policy.linkerd.io";
+const POLICY_API_VERSION: &str = "policy.linkerd.io/v1alpha1";
+const STATUS_CONTROLLER_NAME: &str = "policy.linkerd.io/status-controller";
+
+pub type SharedIndex = Arc<RwLock<Index>>;
+
+pub struct Controller {
+    client: k8s::Client,
+    updates: UnboundedReceiver<Update>,
+}
+
+pub struct Index {
+    updates: UnboundedSender<Update>,
+
+    http_routes: HashMap<ResourceId, RouteBinding>,
+    servers: HashSet<ResourceId>,
+}
+
+pub struct Update {
+    id: ResourceId,
+    patch: k8s::Patch<serde_json::Value>,
+}
+
+impl Controller {
+    pub fn new(client: k8s::Client, updates: UnboundedReceiver<Update>) -> Self {
+        Self { client, updates }
+    }
+
+    pub async fn process_updates(mut self) {
+        let patch_params = k8s::PatchParams::apply("policy.linkerd.io");
+
+        // todo: If an update fails we should figure out a requeueing strategy
+        while let Some(Update { id, patch }) = self.updates.recv().await {
+            let api =
+                k8s::Api::<k8s::policy::HttpRoute>::namespaced(self.client.clone(), &id.namespace);
+
+            // todo: Do we need to consider a timeout here?
+            if let Err(error) = api.patch_status(&id.name, &patch_params, &patch).await {
```

Comment:
`await`ing this means we won't try to patch the next update until this one completes. is that the correct behavior here? 

do the updates need to be applied in order, or do we just want to apply them as quickly as possible? if the intent is to apply them in order, the current approach seems correct, but if we don't need to do this in order, it might be preferable to just spawn a new task with the `patch_status` call, so that we don't need to wait for one API call to finish before we handle the next update...

either way, it might be nice to document whether or not these are order dependent?

Reply:
Yea applying patches in order feels like the right thing to do for this initial implementation. As the follow-ups come in on this and we consider retries, timeouts, and the lease API, I think we get a better idea on if this still needs to be the case.